### PR TITLE
Constraint/required validation UI

### LINKED ISSUE #130: Display constraint and required validation errors
**User story**

As a user filling out a web form,
I want to know when I've missed a question or entered data incorrectly 
so that I can submit my survey without errors

Given ready to submit my long survey (and I have errors)
when click the send button
then I will be taken up to the top of the form and presented with the number of errors with a button to fix the first one

Given I have entered invalid information in a question 
when I proceed to the next question
then I'll see the feedback indicating what went wrong 

Given I have entered invalid information 
when I proceed to send my form
then I'll see the field error feedback indicating what went wrong with the section highlighted 
and when I address the errors, the error count should change accordingly

[Figma
](https://www.figma.com/design/KFi9hIQdrRgo0rLqIC7Sg8/ODK-user-flows?node-id=5570-99332&t=Wn3AoxT4Cu9zAc75-1)

### PR DESCRIPTION
**Pull request by sadiqkhoja**
Closes #130 

### Changes:
- Added new icons for error in icomoon font
- Added violation message for leaf nodes, message is shown only when the field value is changed or 'send' button is pressed. I add `dirty` class to the control when its value is changed. 
- Added Form validation banner at the top, it is shown only when 'send' button is pressed. It has 'fix error' link that scrolls to the first invalid question.

### CODE REVIEW

**Review by eyelidlessness — CHANGES_REQUESTED** [review]
Great to see this in action! Overall it's a great first step on one of the remaining bits of essential functionality for many use cases.

I've added a fair bit of commentary, much of it around approach. While I'd like to spend some time thinking more deeply about the questions I've raised in those comments, I'd generally prioritize them lower (and happy to defer to future work) than a few areas of UX which feel need some improvement.

To my mind, I'd prioritize those (highest to lowest):

1. Reducing the difference between conditions for first- and subsequent-display of input validation messages. I'm not totally sure the suggestion I made is the best/only way to achieve that, but my experience before/after the suggested change is that it's drastically easier to understand what happens in various cases I've tried (both in terms of initial form definition, and various usage and state flows).
2. Some refinement of the form-wide sticky banner for narrow screens. As I mentioned inline, a single-breakpoint to switch from 100% to 70% width would be a great start. I'm happy to consider more specific improvements as well.
3. I do think it'd be good now to anticipate app translations, with some preliminary structure for referencing static/templated user-facing text. This could be as simple as importing those from a common module.

I have several concerns about scrolling (and DOM scroll APIs) as a mechanism for in-form navigation. I think the most immediately pressing aspect there is also the one with the most immediate UX impact: focus. I'd be happy to defer the rest of the scrolling related concerns, but it's worth thinking about whether there's a quick win in terms of _jump to first validation error_ -> _focus field with first validation error_. I think the rest may come into play more as we address other view modes, other aspects of in-form navigation, etc. But if we can reasonably bring the issue of focus into scope, it'll be good for a variety of accessibility/usability reasons.

File: packages/web-forms/src/OdkWebFormDemo.vue
Comment (sadiqkhoja) [reply packages/web-forms/src/OdkWebFormDemo.vue:36]: Point noted. I am thinking that in few week we would have a separate "standalone web-forms service" and that point we can delete this demo app altogether.

File: packages/web-forms/src/components/appearances/FieldListTable.vue
Comment (sadiqkhoja) [reply packages/web-forms/src/components/appearances/FieldListTable.vue]: With the validation placeholder:
image

Without placeholder:
image

I am adding the Form in the repository that we use for creating demo videos.

File: packages/web-forms/src/components/FormQuestion.vue
Comment (sadiqkhoja) [reply packages/web-forms/src/components/FormQuestion.vue]: I agree with you. I was on the edge about where the validation-message element should go. Now I have created a separate component and a special treatment of it for `InputText.vue` goes in its right place (I think so).

File: packages/web-forms/src/components/FormQuestion.vue
Comment (sadiqkhoja) [reply packages/web-forms/src/components/FormQuestion.vue]: this is now removed

File: packages/web-forms/src/components/controls/InputText.vue
Comment (sadiqkhoja) [reply packages/web-forms/src/components/controls/InputText.vue]: I am also liking the experience with `input` and `blur` events.

File: packages/web-forms/src/components/controls/InputText.vue
Comment (sadiqkhoja) [reply packages/web-forms/src/components/controls/InputText.vue]: using `.icon-error` to select the icon for now.

File: packages/web-forms/src/components/controls/InputText.vue
Comment (sadiqkhoja) [reply packages/web-forms/src/components/controls/InputText.vue]: replaced with `inset-inline-end`

File: packages/web-forms/src/components/FormHeader.vue
Comment (sadiqkhoja) [reply packages/web-forms/src/components/FormHeader.vue]: changed the width to 100% for `$lg` breakpoint.

File: packages/web-forms/src/components/FormHeader.vue
Comment (sadiqkhoja) [reply packages/web-forms/src/components/FormHeader.vue:45]: - Added a wrapper div for desktop branch, previously `hidden lg:*` was added to multiple elements for desktop.
- A wrapper div was added in smaller-screen section which is now removed. It was not needed any more after moving `formErrorMessage (PrimeMessage)` to OdkWebForm component

File: packages/web-forms/src/components/widgets/CheckboxWidget.vue
Comment (sadiqkhoja) [reply packages/web-forms/src/components/widgets/CheckboxWidget.vue:52]: I will handle this with https://github.com/getodk/web-forms/pull/173#discussion_r1690526886

File: packages/web-forms/src/components/FormQuestion.vue
Comment (sadiqkhoja) [reply packages/web-forms/src/components/FormQuestion.vue:38]: changed it to `4rem`

File: packages/web-forms/src/components/FormHeader.vue
Comment (sadiqkhoja) [reply packages/web-forms/src/components/FormHeader.vue]: reverted it.

File: packages/web-forms/src/components/OdkWebForm.vue
Comment (sadiqkhoja) [reply packages/web-forms/src/components/OdkWebForm.vue]: Talked with Aly, she liked the suggestion of scrolling to and focusing first invalid question. Made that change.

Still not sure what you are suggesting with "is a DOM scrolling API the way we want to handle programmatic in-form navigation?"

File: packages/web-forms/src/components/OdkWebForm.vue
Comment (eyelidlessness) [inline packages/web-forms/src/components/OdkWebForm.vue:58]: For previous discussion context: [this thread](https://github.com/getodk/web-forms/pull/164#discussion_r1693632522) and [this other one](https://github.com/getodk/web-forms/pull/164#discussion_r1697520738).

> Still not sure what you are suggesting with "is a DOM scrolling API the way we want to handle programmatic in-form navigation?"

I wasn't suggesting anything in particular, it was an intentionally open-ended question. I was hoping to prompt discussion without biasing or coloring the discussion with any alternative details I might have in min. I was hoping that you would give some thought (if you hadn't already) to alternatives that you might find appealing.

But since you've asked for suggestions, here's a rough sketch of an approach I think might be better, both now (for usability and maintainability) and into known future work (where we know that not all questions will always be visible, and so won't necessarily be scroll targets in the first place):

1. Some form of shared/app global state indicating the question being targeted in a way that's generic enough it can be adapted to arbitrary question types. For instance, it could be as simple as setting the state to the targeted question's `nodeId`.

2. At the individual question level, some mechanism to react to that state change.

3. At the targeted question, some consistent way to instruct whatever component renders _its control_ that it should be activated (where "activated" brings it into view, performs whatever pertinent focus behavior, whatever other accessibility affordances we want to consider). This is something that will vary by control type, so it makes sense for _the control itself_ to handle it.

4. _Probably_ some way of making this state ephemeral, so that subsequent user interactions take precedence after the activation occurs.

5. Some way to signal to non-root ancestors when a descendant question is targeted. I'm not sure whether it makes sense for this communication to descend or ascend the hierarchy, but it will be important for paginated views. This is a good candidate to defer to later work, but it's something I want to consciously anticipate in the approach because we already know it will be coming.

Maybe we can do a pairing session to explore this? It seems likely there are some Vue APIs and common idioms that we could reach for. In other component libraries (React, Solid), I'd probably be thinking about utilizing their Context API or equivalent for coordination of global/question-local state, and then I'd probably be thinking about a base interface or "higher order component" for communication between the question and its control.

- - -

Since point 3 mentions that the activation behavior will likely be control-specific, I think it's worth mentioning that the special casing of text inputs here (a) hints at the same concern (glad we're both thinking about it!) but (b) confuses me as implemented (why wouldn't we want to focus other controls?). I'm also anticipating this will balloon into a lot of special cases that become harder to reason about and keep track of.

In general, I'm really nervous about the amount of implicit coupling this introduces now, and the ways that will be harder to manage over time.

As I said earlier, I did intentionally try to avoid making specific suggestions of alternatives. So I hope the rough sketch above won't be taken as too proscriptive about what alternatives might look like. And in that spirit, I'm happy to hear any other approaches you might have in mind as well.

- - -

More broadly, and maybe we can discuss this as a team tomorrow: it's worth thinking about whether we'd benefit from putting more design/discussion of approach upfront before diving into implementation. I think that's worked very well for the engine work we've landed over the last few months, and I think it would benefit the UI side as well. I also know it doesn't feel great to make progress on a big chunk of work, only to have significant questions about its approach come up in hindsight. Having more of a shared understanding of the expected design/approach would certainly help to reduce if not avoid that entirely.

**Review by eyelidlessness — CHANGES_REQUESTED** [review]
One thing I didn't notice on the first pass (sorry!) but jumped out at me now looking at the test changes... it really feels like there's a fair amount of testable logic going on here. It would be nice to have some tests for:

- Question-level messages across at least these types (where and how they differ):
  - String fields
  - Select fields
  - Field/table list

... applied across...

- Message display behavior:
  - Show on/after first interaction setting invalid values
  - Not show on/after first interaction setting valid values
  - Subsequent interaction reverting the first interaction state

- Form-level (submit) behavior:
  - Presence of global banner
  - Some stable check for highlight of invalid questions
  - Behavior associated with resolution of all violations

File: packages/web-forms/src/components/FormHeader.vue
Comment (sadiqkhoja) [reply packages/web-forms/src/components/FormHeader.vue]: Following is the visualization we want to achieve:

image

I had tried using top-layer and pop-over and I could get this:

image

I can certainly add static margin-top in the bottom-layer, but that won't look clean solution.

File: packages/web-forms/src/components/ValidationMessage.vue
Comment (sadiqkhoja) [reply packages/web-forms/src/components/ValidationMessage.vue:24]: this has been fixed in https://github.com/getodk/web-forms/pull/173/commits/8382ffa83987656a4f0bf74af5fd4a5b0664d40a of https://github.com/getodk/web-forms/pull/173

Comment (eyelidlessness) [reply packages/web-forms/src/components/ValidationMessage.vue:24]: One of the challenges with stacking branches/PRs is addressing issues found in an earlier one, and propagating that out to later ones. The approach that has generally kept @lognaturel happy when she's reviewing my stacked work is addressing feedback on the branch introducing the issue, and then rebasing downstream as needed. If you don't want to do that here, given there are at least a couple of these comments saying feedback is addressed in the downstream branch, maybe we should consider whether it'd be more effective as a single PR? That's not my favorite option, but I'm trying to be flexible to help move this forward.

File: packages/web-forms/src/components/FormHeader.vue
Comment (eyelidlessness) [reply packages/web-forms/src/components/FormHeader.vue]: Marking this thread resolved, hopefully to consolidate discussion in https://github.com/getodk/web-forms/pull/164#discussion_r1714415457.

File: packages/web-forms/src/components/FormHeader.vue
Comment (eyelidlessness) [reply packages/web-forms/src/components/FormHeader.vue]: There is already one other place I can think of that applies a style specifically reserving space for this banner. Any reason it couldn't be extracted to a CSS variable, referenced in both/all appropriate places?

File: packages/web-forms/src/components/OdkWebForm.vue
Comment (eyelidlessness) [reply packages/web-forms/src/components/OdkWebForm.vue]: As with the other outdated thread on the topic, I'll mark this one resolved so we can consolidate discussion of the topic here https://github.com/getodk/web-forms/pull/164#discussion_r1714415457

File: packages/web-forms/src/components/appearances/FieldListTable.vue
Comment (eyelidlessness) [reply packages/web-forms/src/components/appearances/FieldListTable.vue]: Thanks! This is helpful clarification. I'll resolve here for now. I do want to mention that when I was experimenting with the message styling, I noticed it was easy to introduce a vertical layout shift between rows when selecting an item. I don't see that on this branch, but I _did see it_ in the PrimeVue 4 demo linked yesterday. Something to keep an eye out for.

File: packages/web-forms/src/components/FormQuestion.vue
Comment (eyelidlessness) [reply packages/web-forms/src/components/FormQuestion.vue]: Also resolving this to consolidate remaining discussion of scrolling for navigation at https://github.com/getodk/web-forms/pull/164#discussion_r1714415457

File: packages/web-forms/src/components/OdkWebForm.vue
Comment (sadiqkhoja) [reply packages/web-forms/src/components/OdkWebForm.vue]: Removed "Fix errors" button and created https://github.com/getodk/web-forms/issues/189 to address that.

File: packages/web-forms/src/components/FormHeader.vue
Comment (sadiqkhoja) [reply packages/web-forms/src/components/FormHeader.vue]: removed z-index and used popover

File: packages/ui-solid/fixtures/xforms/socio-economic-survey.xml
Comment (sadiqkhoja) [reply packages/ui-solid/fixtures/xforms/socio-economic-survey.xml]: fixed.

File: packages/web-forms/src/components/ValidationMessage.vue
Comment (sadiqkhoja) [reply packages/web-forms/src/components/ValidationMessage.vue]: merged #173  into this.

File: packages/web-forms/tests/components/OdkWebForm.test.ts
Comment (eyelidlessness) [inline packages/web-forms/tests/components/OdkWebForm.test.ts:33]: I know there's a lot here, but it's worth considering. There are a few good candidates for moving to somewhere shared/reusable, but we can start here for now.

Main goals:

1. Clean up any mutation of global scope/prototypes
2. Use mocking facilities consistently
3. Generalize enough that we can apply similar techniques without a lot of extra fuss in the future

```suggestion
	// TODO: how the heck is `undefined` a key of anything?!
	type StringKeyOf<T> = Extract<keyof T, string>;

	type StringKeyOfDocument = StringKeyOf<Document>;

	// prettier-ignore
	type DocumentPropertyName = {
		[PropertyName in StringKeyOfDocument]:
			Document[PropertyName] extends AnyFunction
				? never
				: PropertyName;
	}[StringKeyOfDocument];

	let documentKeysAdded: StringKeyOfDocument[] = [];

	const mockDocumentGetter = <PropertyName extends DocumentPropertyName>(
		propertyName: PropertyName,
		mockImplementation: () => Document[PropertyName]
	) => {
		if (propertyName in document) {
			return vi.spyOn(document, propertyName, 'get').mockImplementation(mockImplementation);
		}

		documentKeysAdded.push(propertyName);

		const mock = vi.fn(mockImplementation);

		Object.defineProperty(document, propertyName, {
			get: mock,
			configurable: true,
		});

		return mock;
	};

	type StringKeyOfHTMLElement = StringKeyOf<HTMLElement>;

	// prettier-ignore
	type ElementMethodName = {
		[PropertyName in StringKeyOfHTMLElement]:
			HTMLElement[PropertyName] extends AnyFunction
				? PropertyName
				: never;
	}[keyof HTMLElement];

	type ElementMethodMock<MethodName extends ElementMethodName> = (
		this: HTMLElement,
		...args: Parameters<HTMLElement[MethodName]>
	) => ReturnType<HTMLElement[MethodName]>;

	let elementKeysAdded: StringKeyOfHTMLElement[] = [];

	const mockElementPrototypeMethod = <MethodName extends ElementMethodName>(
		methodName: MethodName,
		mockImplementation: ElementMethodMock<MethodName>
	) => {
		if (methodName in HTMLElement.prototype) {
			return vi.spyOn(HTMLElement.prototype, methodName).mockImplementation(function (...args) {
				// eslint-disable-next-line @typescript-eslint/no-unsafe-argument, @typescript-eslint/no-explicit-any
				return mockImplementation.apply(this as HTMLElement, args as any);
			});
		}

		elementKeysAdded.push(methodName);

		const mock = vi.fn(mockImplementation);

		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-explicit-any
		HTMLElement.prototype[methodName] = mock as any;

		return mock;
	};

	beforeEach(() => {
		documentKeysAdded = [];
		elementKeysAdded = [];

		// Not necessary if we scroll `document.scrollingElement` instead.
		// Including to show how it'd be done if not.
		vi.stubGlobal('scrollTo', () => {
			/* Do nothing, unless we test this behavior. */
		});

		mockDocumentGetter('scrollingElement', () => {
			return document.documentElement;
		});

		mockElementPrototypeMethod('showPopover', function () {
			this.style.display = 'block';
		});

		mockElementPrototypeMethod('hidePopover', function () {
			this.style.display = 'none';
		});
	});

	afterEach(() => {
		documentKeysAdded.forEach((propertyName) => {
			delete document[propertyName];
		});

		elementKeysAdded.forEach((methodName) => {
			delete HTMLElement.prototype[methodName];
		});

		vi.restoreAllMocks();
		vi.unstubAllGlobals();
	});
```

(You'll also need these import changes:)

```diff
- import { describe, expect, it } from 'vitest';
+ import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
+ import type { AnyFunction } from '@getodk/common/types/helpers.d.ts';
```

File: packages/web-forms/tests/components/OdkWebForm.test.ts
Comment (eyelidlessness) [inline packages/web-forms/tests/components/OdkWebForm.test.ts:63]: This is at least 2 tests. Fine for now, but probably worth breaking up when we revisit the programmatic in-form navigation aspect.

**Review by eyelidlessness — CHANGES_REQUESTED** [review]
Getting really close! I left a few suggestions inline. Feel free to take them or leave them now, but I'd strongly encourage the ones related to scrolling and mocking.

I **would like** to see a changeset before merge. I'd be happy to add these for you when they're missing, if that's what you'd prefer. Otherwise I'll suggest that I've found the GitHub bot provides a helpful reminder.

These are the aspects I think are worth addressing or considering now.

I'll add a separate comment with a variety of other notes I took during this review pass, which we can file in one or more issues for followup as appropriate.

**Comment by eyelidlessness** [comment]
(Sorry if any formatting is off here, or anything is confusing. These were notes I was scrawling in Notes.app while I reviewed, and may not have clarified or Markdown-ified everything sufficiently.)

### Nice to have in follow-up PR(s)

#### Vue conventions

* When using `provide`/`inject`, would be nice to establish a pattern where they use normal references so the connection between provided/injected state can be followed in editor, and types can be defined in provide. I did a quick spike on this not long ago and it seemed possible, happy to dig up the stash if there’s interest.
* Events make me nervous for reasons similar to provide/inject. They seem a bit easier to follow at component boundaries, but the lack of clear types with a single source of truth is still concerning.
* Consider some DOM-level way of identifying which component rendered what to the DOM
* Maybe there’s a lint rule (or maybe we could write one trivially?) to ensure consistent use of `class`/`:class`? It’s confusing every time I see both used on the same element. (And maybe I’ll get over that seeing it more?)
* Consider extracting event handlers into functions, especially where their logic is repeated. Case that stood out/made me think about this: several additions of `@change="touched = true”`.
* How do we choose between when to apply utility classes, versus using styles that achieve the same? (flex etc stand out) 
* The tooling boundary between .vue and .ts modules is a continuing headache. Common case: TS/ESLint differing on whether a type is `any`. Case that stood out here: props interface exported from component, referenced in test. Would like to consider other approaches that minimize this. And/or set aside some time to see if any tooling updates address these shortcomings.
* I’m normally not fond of “test helpers”, but have some concern about inconsistent use of `mount` and `globalMountOptions`: especially (but not only) whether or not PrimeVue plugin is in use. Worth considering a shared custom `mount` helper to ensure we’re testing all components under the same assumptions.

#### Validation

* It would be good if select constraint messages don’t show up on first selection if user hasn’t moved on. Inputs use blur. It would require a little more logic to produce a similar experience for selects as radios/checkboxes (e.g. some delay on blur to cancel on interaction with additional items)

#### Styles

* Language menu in wide view should be its widest available width and keep a consistent width when switching languages
* Needs better engine support: would be nice to not show validation placeholder where a field will never show a validation message (i.e. has neither a `constraint` or `required` expression)
* Some way of visually distinguishing UnsupportedAppearance (and similar dev-focused components, maybe worth having a base component for them to compose)
* FormHeader.vue:
    * `.smaller-screens .print-button` doesn’t render justified end when form has no translations
    * Consider changing toggle of language menu visibility to `languages.length > 1`. (It’s not common, but I’ve seen some forms define only a single language; also worth considering suppressing languages in the engine for this scenario instead) 
    * `.smaller-screens` shadow looks larger/heavier than design. Much closer if we replace the current `filter` rule with:

        ```css
        position: relative; /* to ensure shadow is visible */
        box-shadow: 0px 2px 6px 0px rgba(0, 0, 0, 0.15), 0px 1px 2px 0px rgba(0, 0, 0, 0.3);
        ```

        This does differ from the styles produced by Figma. I suspect we’ll need similar finesse/eyeball comparison for other styles we derive from the designs directly
* In #173, made comment about leaning more on flex/grid for alignment instead of small pixel/small unit alignment adjustments
* Naming things: maybe this will be better addressed with PrimeVue 4, but it would be nice if we had fewer magic values and more use of system of CSS variables or other named values. Will hopefully make it easier to understand why something changes, e.g. font-weight 400 to 500 (or vice versa), font-size changes, accommodations for spacing moving up or down component/DOM hierarchy
* Various use of negative margins is concerning. Especially where it seems to be counteracting spacing controlled by the same general thing (`FormPanel`). Would generally like to do a focused overhaul on how layout/spacing/alignment/etc are handled.
    * In such focused overhaul, also want to be mindful of RTL where we have styles producing asymmetric horizontal spacing/alignment
* There’s a layout shift when validation message is shown/hidden within rows of FieldListTable

#### Component structure/naming

* (Not directly related to PR, want to track as it’s come up in recent conversation) FieldListTable naming is confusing. I think we’ve both repeatedly associated it with the `field-list` appearance, but it is not.
* In general, worth thinking about:
    * System for naming and categorical structuring of components
    * How that can/will align with designs in Figma, tokens in PrimeVue, our own token system
* Adding the same thing in multiple places can be a smell, may suggest there’s a composable component wanting to exist. Case that stood out/made me think about this: several additions of `@change="touched = true”`.
* Props naming:
    * “add” is a weird prefix for something that’s not an action/handler/callback. In general, verbs tend to suggest some kind of event handling or interaction -> state reaction.
* Events naming: we should consider developing a predictable system for this as well

#### Interactivity

* A little surprised by DOM API usage for `popover`, but didn’t realize PrimeVue’s Popover component was introduced in v4. We should track migration to that component in the v4 transition, if it’s suitable. If not, worth considering a composable base component to isolate the DOM API aspects in an obvious place

**Review by eyelidlessness — APPROVED** [review]

**Pull request merged**
